text/2018-10-25-batch-split.md

```diff
@@ -0,0 +1,170 @@
+}
+
+message BatchSplitResponse {
+    repeated metapb.Region regions = 1;
+}
+
+enum AdminCmdType {
+    // ...
+    Split = 2 [deprecated=true];
+    // ...
+    BatchSplit = 10;
+}
+
+message AdminRequest {
+    // ...
+    SplitRequest split = 3 [deprecated=true];
+    // ...
+    BatchSplitRequest splits = 10;
+}
+
+message AdminResponse {
+    // ...
+    SplitResponse split = 3 [deprecated=true];
+    // ...
+    BatchSplitResponse splits = 10;
+}
+```
+
+Add a new admin command type `BatchSplit`  with related request and response. `BatchSplitRequest` wraps multiple `SplitRequest` along with `right_derive` which invalidates the `right_derive` in each `SplitRequest`.
+
+When in rolling upgrade process, new TiKVs are mixed up with old TiKVs, so old command type `Split` still needs to be preserved.
+
+## Implementation in TiKV
+
+### How to produce multiple split keys
+
+This part mainly focus on `SplitChecker`. 
+
+First of all, adjust trait to make it can return multiple split keys.
+
+```rust
+pub trait SplitChecker {
+    // ...
+    
+    // before: fn split_key(&mut self) -> Option<Vec<u8>>
+    fn split_keys(&mut self) -> Vec<Vec<u8>>;
+    
+    // before: fn approximate_split_key(&self, _: &Region, _: &DB) -> Result<Option<Vec<u8>>> 
+    fn approximate_split_keys(&self, _: &Region, _: &DB) -> Result<Vec<Vec<u8>>> {
+}
+```
+
+Then add one config `batch_split_limit` to limit the number of produced split keys in a batch. If it is unlimited, for once split check, it scans all over the Region's range, and in some extreme case this would cause performance issue.
+
+Now we have four split-checkers: half, keys, size and table. SizeChecker and KeysChecker can be rewritten to produce multiple keys, and other checkers' logic stay unchanged. 
+
+The general logic of SizeChecker and KeysChecker are similiar, the only difference between them is one splits Region based on size and the other splits Region based on the number of keys. So here we mainly describe the logic of SizeChecker:
+
+- before: it scans key-value pairs in a Region's range sequentially to accumlate their size as `total_size` and stops once the size reachs to `region_max_size` or scans to the end of range. If  `total_size` is smaller than `region_max_size` at the end, checker wouldn't produce any split key; if not, it regards the very key at which `total_size`  reachs to `region_split_size` as split key.
+- after: it scans key-value pairs in a Region's range sequentially to accumlate their size as `total_size` and stops once the size reachs to `region_split_size * (batch_split_limit-1) + region_max_size` or scans to the end of range. During the scan process, it reocrds the key as split key every `region_split_size`, but after finishing scanning, it may discards the last split key if the size of rest Region doesn't over `region_max_size - region_split_size`. With this algorithm, if `batch_split_limit` is set to 1, TiKV can perfectly behave as before that split without batch.
```

Comment:
```suggestion
- after: it scans key-value pairs in a Region's range sequentially to accumulate their size as `total_size` and stops once the size reaches `region_split_size * (batch_split_limit-1) + region_max_size` or scans to the end of the range. During the scan process, it records the key as split key every `region_split_size`, but after finishing scanning, it may discard the last split key if the size of rest Region is not bigger than `region_max_size - region_split_size`. With this algorithm, if `batch_split_limit` is set to 1, TiKV can perfectly behave the same way as the split without batch.
```